src/lib/ares_process.c

```diff
@@ -79,26 +114,29 @@ static void          server_increment_failures(struct server_state *server)
   if (node == NULL) {
     return;
   }
+
   server->consec_failures++;
   ares__slist_node_reinsert(node);
+
+  invoke_server_state_cb(server, ARES_FALSE);
```

Comment:
I'd think this should only be called on the first failure, not every.

Reply:
So I think there are use-cases for calling on every success/failure. It would give extra control to applications, although I can understand that calling for every query (possibly multiple times) is a lot of overhead.

E.g. if an application wants to track metrics for how many success/failed requests are sent to each server.

Also, in my case knowing if a server is failing repeatedly rather than just a one-off is useful, to detect a more serious error rather than a transient network issue.

We should make a decision between:
- Calling on all failures and successes.
- Calling on just the first failure and the first success (that changes the state).

The former gives more control, whilst the latter is a simpler interface to use. Do you have a feel for which would be useful more generally?

Reply:
Also, if we go for the former then I think we should cache the string representation of the server in the `server_state` object; having a pre-built string that we can return a `const` pointer to would save on performance.

Reply:
You're the one with the actual usecase for this, are you taking some 'action' immediately when notified of a failure or anything like that?  I sort of wonder if it would make sense to have a server status query, where it would populate a data structure with all servers and statistics that could be iterated across so you don't have to actually track each server's state.